fix: multiple widgets for `Booker` atom

sean-brydon opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes an issue in the `Booker` atom wherein previously we were not able to use multiple instance of the `Booker` widget on the same page. If we would use multiple `Booker` widgets the state would change in both of them even though one of them would be toggled
- This fix ensures each instance of the `Booker` atom widget uses a store of its own instead of sharing a global instance that all the widgets can access which was the previous approach

## Visual Demo:

https://github.com/user-attachments/assets/7729888c-0d3b-4d90-a8e5-ead047a8f0f6


## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] (N/A) I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] (N/A) I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
For the web app these are the steps:
1. Run the web app in local
2. Go to the public profile page of user and navigate to one of his events and try to book them via the booker
3. There should not be any change in the behaviour of the Booker widget for the web app so ideally it should stay the same


For platform these are the steps:
1. Run the examples app we have [here](https://github.com/calcom/cal.com/tree/main/packages/platform/examples/base)
2. Go to `booking.tsx` in the examples app and copy the code for Booker atom [here](https://github.com/calcom/cal.com/blob/main/packages/platform/examples/base/src/pages/booking.tsx#L93) and paste it just under to make another instance of the same `Booker` widget appear  
3. In both of the `Booker` atom widgets you need to either comment out the prop allowUpdatingUrlParams or set it to false
4. Try to change something in one atom and see if the state of the other widget changes or not

sean-brydon committed:
fix manage test error

sean-brydon committed:
fix manage test error

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

Ryukemeister committed:
update useBookerContext

Ryukemeister committed:
wrap store provider around booker

Ryukemeister committed:
replace useStore with useBookerStoreContext

Ryukemeister committed:
replace useBookerStore with useBookerStoreContext

Ryukemeister committed:
add initializer function for booker store provider

Ryukemeister committed:
update props

Ryukemeister committed:
fixup: pass more props

Ryukemeister committed:
export StoreInitializeType type

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

Ryukemeister committed:
replace useBookerStore with useBookerStoreContext

Ryukemeister committed:
fixup: dont wrap BookerStoreProvider around booker directly

Ryukemeister committed:
wrap BookerStoreProvider around booker web wrapper and fix local storage import

devin-ai-integration[bot] committed:
fix: wrap test components with BookerStoreProvider to fix failing unit tests

- Create reusable test utility in test-utils.tsx with comprehensive mock store
- Update Booker.test.tsx to use context-based testing approach
- Fix DatePicker tests in both bookings and calendars packages
- Simulate auto-advance behavior for month navigation tests
- All 14 previously failing tests now pass

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

Ryukemeister committed:
add changesets

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

Ryukemeister reviewed (commented):
TLDR here's a short summary of the changes made in this PR
1. `useBookerStore` is being replaced by a new context called `useBookerStoreContext`, where each Booker widget has its own store instead of sharing the same global store instance from `useBookerStore`
2. hence a lot of variables now fetch values from `useBookerStoreContext` hook
3. in a couple of components some values or event handlers are now being accepted as props instead of fetching their values, reason behind it is that we use the same values in the parent component so the child component just accepts the value instead of calling `useBookerStoreContext` again
4. however there should be no change in the behaviour of the `Booker` in the web app, this should only affect the `Booker` atom which we have in platform

supalarry reviewed (dismissed):
Amazing work Rajiv!

hbjORbj committed:
Merge branch 'main' into rajiv-demo

hbjORbj reviewed (dismissed):
Can you check on the failing tests?

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

devin-ai-integration[bot] committed:
fix: complete migration from useBookerStore to useBookerStoreContext

- Updated all remaining files to use useBookerStoreContext instead of useBookerStore
- Fixed BookingFields.test.tsx by wrapping test component with BookerStoreProvider
- Ensures all Booker components work within the new context-based store pattern
- Resolves failing unit tests in CI by completing the store migration

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
fix: add missing useBookerStoreContext import in BookerWebWrapper.tsx

- Fixes TypeScript errors in CI where useBookerStoreContext was used but not imported
- Completes the migration from useBookerStore to useBookerStoreContext pattern
- All tests pass locally after this fix

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
Revert "fix: add missing useBookerStoreContext import in BookerWebWrapper.tsx"

This reverts commit 12947ca63f46c42e5aee74c78176ef230e2bf2cf.

devin-ai-integration[bot] committed:
Revert "fix: complete migration from useBookerStore to useBookerStoreContext"

This reverts commit 87d837f2d30e4f9259d2be786049d21856f4eef0.

Ryukemeister committed:
fixup: useBookerStoreContext and useInitializeBookerStoreContext for booker web wrapper

Ryukemeister committed:
fixup: failing E2E tests

Ryukemeister committed:
fixup: email embed breaking

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

Ryukemeister committed:
fixup: troubleshooter crashing

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

Ryukemeister committed:
replace useBookerstore with useBookerStoreContext

Ryukemeister committed:
Merge branch 'main' into rajiv-demo

CarinaWolli reviewed (approved):
Tested the web app booker, and everything is still working as it should
